alibi_detect/od/_svm.py

```diff
@@ -0,0 +1,233 @@
+            Device type used. The default tries to use the GPU and falls back on CPU if needed. Can be specified by
+            passing either ``'cuda'``, ``'gpu'``, ``'cpu'`` or an instance of ``torch.device``.
+
+        Raises
+        ------
+        NotImplementedError
+            If choice of `backend` is not implemented.
+        ValueError
+            If choice of `optimization` is not valid.
+        ValueError
+            If `n_components` is not a positive integer.
+        """
+        super().__init__()
+
+        if optimization not in ('sgd', 'gd'):
+            raise ValueError(f'Optimization {optimization} not recognized. Choose from `sgd` or `gd`.')
+
+        if n_components is not None and n_components <= 0:
+            raise ValueError(f'n_components must be a positive integer, got {n_components}.')
+
+        backend_str: str = backend.lower()
+        BackendValidator(
+            backend_options={'pytorch': ['pytorch']},
+            construct_name=self.__class__.__name__
+        ).verify_backend(backend_str)
+
+        backend_cls = backends[backend][optimization]
+        args: Dict[str, Any] = {
+            'n_components': n_components,
+            'kernel': kernel,
+            'nu': nu
+        }
+        args['device'] = device
+        self.backend = backend_cls(**args)
+
+    def fit(
+        self,
+        x_ref: np.ndarray,
+        tol: float = 1e-6,
+        max_iter: int = 1000,
+        step_size_range: Tuple[float, float] = (1e-6, 1.0),
+        n_step_sizes: int = 16,
+        n_iter_no_change: int = 25,
+        verbose: int = 0,
+    ) -> None:
+        """Fit the detector on reference data.
+
+        Uses the choice of optimization method to fit the svm model to the data.
+
+        Parameters
+        ----------
+        x_ref
+            Reference data used to fit the detector.
+        tol
+            Convergence threshold used to fit the detector. Used for both ``'sgd'`` and ``'gd'`` optimizations.
+            Defaults to ``1e-3``.
+        max_iter
+            The maximum number of optimization steps. Used for both ``'sgd'`` and ``'gd'`` optimizations.
+        step_size_range
+            The range of values to be considered for the gradient descent step size at each iteration. This is specified
+            as a tuple of the form `(min_eta, max_eta)` and only used for the ``'gd'`` optimization.
+        n_step_sizes
+            The number of step sizes in the defined range to be tested for loss reduction. This many points are spaced
+            equidistantly along the range in log space. This is only used for the ``'gd'`` optimization.
+        n_iter_no_change
+            The number of iterations over which the loss must decrease by `tol` in order for optimization to continue.
+            This is only used for the ``'gd'`` optimization..
+        verbose
+            Verbosity level during training. ``0`` is silent, ``1`` a progress bar or sklearn training output for the
```

Comment:
`verbose` description is referring to `sklearn` implementation only. Unclear if `verbose` is still relevant when the `torch` backend is used. Would be good to make the description more generic, or be more descriptive as to when the `verbose` kwarg is relevant.